File: composeApp/src/androidMain/kotlin/eu/pitlap/shared/standings/StandingsViewModel.kt
package eu.pitlap.shared.standings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import eu.pitlap.shared.modules.Pitlap
import eu.pitlap.shared.modules.PitlapService
import eu.pitlap.shared.standings.models.StandingRowUiModel
import eu.pitlap.shared.standings.state.StandingListEvent
import eu.pitlap.shared.standings.state.StandingListScreenState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class StandingsViewModel(
    private val pitlapService: PitlapService = Pitlap.getService()
) : ViewModel() {

    private val _state = MutableStateFlow(StandingListScreenState())
    val state: StateFlow<StandingListScreenState> = _state.asStateFlow()

    fun onEvent(event: StandingListEvent) {
        when (event) {
            StandingListEvent.LoadDriverStanding -> loadStandings(
                fetch = { pitlapService.getDriverStandings() },
                mapToUi = {
                    StandingRowUiModel(
                        position = it.positionText,
                        fullName = "${it.givenName} ${it.familyName}",
                        constructorName = it.constructorName,
                        points = it.points.toString(),
                        wins = it.wins.toString()
                    )
                }
            )

            StandingListEvent.LoadConstructorStanding -> loadStandings(
                fetch = { pitlapService.getConstructorStandings() },
                mapToUi = {
                    StandingRowUiModel(
                        position = it.positionText,
                        fullName = null,
                        constructorName = it.constructorName,
                        points = it.points.toString(),
                        wins = it.wins.toString()
                    )
                }
            )
        }
    }

    private fun <T> loadStandings(
        fetch: suspend () -> List<T>,
        mapToUi: (T) -> StandingRowUiModel
    ) {
        _state.update { it.copy(isLoading = true, error = null) }

        viewModelScope.launch {
            try {
                val standings = fetch().map(mapToUi)
                _state.update { it.copy(isLoading = false, standings = standings) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }
}

File: iosApp/iosApp/features/standings/viewmodel/StandingsViewModel.swift
//
//  DriverStandingsViewModel.swift
//  PitLap
//
//  Created by Tinashe MAKUTI on 21/01/2025.
//

import Foundation
import Shared

final class StandingsViewModel: ObservableObject {
    private let pitlapService: PitlapService
    
    @Published var isLoading: Bool = false
    @Published var standings: [StandingRowUiModel] = []
    
    init(pitlapService: PitlapService = Pitlap.shared.getService()) {
        self.pitlapService = pitlapService
    }
    
    func getDriverStandings() async {        
        await fetchStandings { [self] in try await pitlapService.getDriverStandings().map { StandingRowUiModel(from: $0) } }
    }
    
    func getConstructorStandings() async {
        await fetchStandings { [self] in try await pitlapService.getConstructorStandings().map { StandingRowUiModel(from: $0) } }
    }
    
    @MainActor
    private func fetchStandings(_ operation: @escaping () async throws -> [StandingRowUiModel]) async {
        isLoading = true
        defer { isLoading = false }
        do {
            self.standings = try await operation()
        } catch {
            print(error.localizedDescription)
        }
    }
}

extension StandingRowUiModel {
    init(from standing: DriverStandingModel) {
        self.init(
            position: standing.positionText,
            fullName: "\(standing.givenName) \(standing.familyName)",
            constructorName: standing.constructorName,
            points: "\(standing.points)",
            wins: "\(standing.wins)"
        )
    }
    
    init(from standing: ConstructorStandingModel) {
        self.init(
            position: standing.positionText,
            fullName: nil,
            constructorName: standing.constructorName,
            points: "\(standing.points)",
            wins: "\(standing.wins)"
        )
    }
}
